introspect/default_inspector.go

```diff
@@ -0,0 +1,240 @@
+package introspect
+
+import (
+	"fmt"
+	"math"
+	"runtime"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/introspection"
+	"github.com/libp2p/go-libp2p-core/introspection/pb"
+	"github.com/libp2p/go-libp2p-core/metrics"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	"github.com/libp2p/go-eventbus"
+	"github.com/multiformats/go-multiaddr"
+)
+
+var _ introspection.Introspector = (*DefaultIntrospector)(nil)
+
+// DefaultIntrospector is an object that introspects the system.
+type DefaultIntrospector struct {
+	*eventManager
+
+	host     host.Host
+	bus      event.Bus
+	wsub     event.Subscription
+	reporter metrics.Reporter
+	started  time.Time
+}
+
+func NewDefaultIntrospector(host host.Host, reporter metrics.Reporter) (introspection.Introspector, error) {
+	bus := host.EventBus()
+	if bus == nil {
+		return nil, fmt.Errorf("introspector requires a host with eventbus capability")
+	}
+
+	sub, err := bus.Subscribe(event.WildcardSubscription, eventbus.BufSize(256))
+	if err != nil {
+		return nil, fmt.Errorf("failed to susbcribe for events with WildcardSubscription")
+	}
+
+	d := &DefaultIntrospector{
+		eventManager: newEventManager(sub.Out()),
+		host:         host,
+		bus:          bus,
+		wsub:         sub,
+		reporter:     reporter,
+		started:      time.Now(),
+	}
+
+	return d, nil
+}
+
+func (d *DefaultIntrospector) Close() error {
+	if err := d.wsub.Close(); err != nil {
+		return fmt.Errorf("failed while trying to close wildcard eventbus subscription: %w", err)
+	}
+
+	close(d.closeCh)
+	d.closeWg.Wait()
+
+	return nil
+}
+
+func (d *DefaultIntrospector) FetchRuntime() (*pb.Runtime, error) {
+	return &pb.Runtime{
+		Implementation: "go-libp2p",
+		Platform:       runtime.GOOS,
+		PeerId:         d.host.ID().Pretty(),
+		Version:        "",
```

Comment:
why is the version empty?

Reply:
We don't really have a go-libp2p version/commit anyway to send back, as far as I'm aware?

Reply:
maybe use the user agent?